e2etest/l2tests/assignment.go

```diff
@@ -678,4 +697,267 @@ var _ = ginkgo.Describe("IP Assignment", func() {
+				&svc1.Status.LoadBalancer.Ingress[1]))
+			Expect(err).NotTo(HaveOccurred())
+		})
+		ginkgo.It("Single-stack pool should assign based on svc changes between PreferDualStack and RequireDualStack", func() {
+			pool1 := metallbv1beta1.IPAddressPool{
+				ObjectMeta: metav1.ObjectMeta{Name: "test-ns-dualstack-pool"},
+				Spec: metallbv1beta1.IPAddressPoolSpec{
+					Addresses: []string{
+						v4PoolAddresses,
+					},
+					AllocateTo: &metallbv1beta1.ServiceAllocation{Namespaces: []string{testNamespace}},
+				},
+			}
+			resources := config.Resources{
+				Pools: []metallbv1beta1.IPAddressPool{pool1},
+			}
+
+			err := ConfigUpdater.Update(resources)
+			Expect(err).NotTo(HaveOccurred())
+
+			ginkgo.By("Creating the service with PreferDualStack policy")
+			svc1, _ := service.CreateWithBackend(cs, testNamespace, "svc-test-ns-pool-1", func(svc *v1.Service) {
+				svc.Spec.IPFamilyPolicy = &IPFamilyPolicyPreferDualStack
+			})
+			Expect(err).NotTo(HaveOccurred())
+
+			ginkgo.By("Validating that 1 IP is assigned to the service")
+			Eventually(func() int {
+				svc1, err = cs.CoreV1().Services(svc1.Namespace).Get(context.Background(), svc1.Name, metav1.GetOptions{})
+				Expect(err).NotTo(HaveOccurred())
+				return len(svc1.Status.LoadBalancer.Ingress)
+			}, 5*time.Minute, 1*time.Second).Should(Equal(1))
+			ginkgo.By("validate LoadBalancer IP is allocated from pool1")
+			err = config.ValidateIPInRange([]metallbv1beta1.IPAddressPool{pool1}, jigservice.GetIngressPoint(
+				&svc1.Status.LoadBalancer.Ingress[0]))
+			Expect(err).NotTo(HaveOccurred())
+			ginkgo.By("Updating svc1 to have RequireDualStack family policy")
+			err = updateServiceIPFamilyPolicy(cs, svc1.Namespace, svc1.Name, IPFamilyPolicyRequireDualStack)
+			Expect(err).NotTo(HaveOccurred())
+
+			ginkgo.By("Validating that the service has no IP")
+			Eventually(func() int {
+				svc1, err = cs.CoreV1().Services(svc1.Namespace).Get(context.Background(), svc1.Name, metav1.GetOptions{})
+				Expect(err).NotTo(HaveOccurred())
+				return len(svc1.Status.LoadBalancer.Ingress)
+			}, 5*time.Minute, 1*time.Second).Should(Equal(0))
+
+			ginkgo.By("Updating svc1 to have PreferDualStack family policy")
+			err = updateServiceIPFamilyPolicy(cs, svc1.Namespace, svc1.Name, IPFamilyPolicyPreferDualStack)
+			Expect(err).NotTo(HaveOccurred())
+			ginkgo.By("Validating that 1 IP is assigned to the service")
+			Eventually(func() int {
+				svc1, err = cs.CoreV1().Services(svc1.Namespace).Get(context.Background(), svc1.Name, metav1.GetOptions{})
+				Expect(err).NotTo(HaveOccurred())
+				return len(svc1.Status.LoadBalancer.Ingress)
+			}, 5*time.Minute, 1*time.Second).Should(Equal(1))
+			ginkgo.By("validate LoadBalancer IP is allocated from pool1")
+			err = config.ValidateIPInRange([]metallbv1beta1.IPAddressPool{pool1}, jigservice.GetIngressPoint(
+				&svc1.Status.LoadBalancer.Ingress[0]))
+			Expect(err).NotTo(HaveOccurred())
+		})
+	})
```

Comment:
can you please add another case for when there is an ipv4 pool only, the service with prefer gets a single stack ipv4, then we add a dualstack pool and verify that the service ip didn't jump to the other pool?